project-backend/Providers/WorkerApplicationProvider/WorkerApplicationProvider.cs

```diff
@@ -0,0 +1,113 @@
+﻿using project_backend.Models.DAOs;
+using project_backend.Models.Exceptions;
+using project_backend.Models.Worker;
+using project_backend.Repos;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace project_backend.Providers.WorkerApplicationProvider
+{
+    public class WorkerApplicationProvider : IWorkerApplicationProvider
+    {
+        private readonly DatabaseContext _dbContext;
+
+        public WorkerApplicationProvider(DatabaseContext databaseContext)
+        {
+            _dbContext = databaseContext;
+        }
+
+
+        public WorkerApplicationDAO AddWorkerApplication(int userId)
+        {
+            var userAlreadyApplied = _dbContext.WorkerApplications.Find(userId);
+            var userIsAlreadyWorker = _dbContext.Workers.Find(userId);
+
+            //Even if the endpoint is guarded by a user role, some users which didn't get their updated token could still call this.
+            if (userIsAlreadyWorker != null)
+            {
+                throw new UserIsAlreadyWorker();
+            }
+
+            if (userAlreadyApplied == null)
+            {
+                var newWorkerApplication = new WorkerApplicationDAO { UserId = userId };
+                _dbContext.WorkerApplications.Add(newWorkerApplication);
+                _dbContext.SaveChanges();
+                return newWorkerApplication;
+            }
+            else
+            {
+                throw new DuplicateWorkerApplicationException();
+            }
+        }
+
+        public void AcceptWorkerApplication(int userId)
+        {
+            var workerApplication = _dbContext.WorkerApplications.Find(userId);
+            var userIsAlreadyWorker = _dbContext.Workers.Any(a => a.UserId == userId);
+
+            if (userIsAlreadyWorker)
+            {
+                throw new UserIsAlreadyWorker();
+            }
+
+            if (workerApplication != null)
+            {
+                var transaction = _dbContext.Database.BeginTransaction();
+
+                _dbContext.WorkerApplications.Remove(workerApplication);
+                _dbContext.SaveChanges();
+
+                var newWorker = new WorkerDAO { UserId = userId };
+                _dbContext.Workers.Add(newWorker);
+                _dbContext.SaveChanges();
+
+                transaction.Commit();
```

Comment:
Hope this bit is ok. Transactions aren't supported by the in-memory DB so I couldn't test it.

Reply:
[Microsoft docs](https://docs.microsoft.com/en-us/ef/core/saving/transactions) say that, by default, all operations done before a SaveChanges call are applied in a single transaction; maybe something like this would be better:
```suggestion
                _dbContext.WorkerApplications.Remove(workerApplication);

                var newWorker = new WorkerDAO { UserId = userId };
                _dbContext.Workers.Add(newWorker);
                
                _dbContext.SaveChanges();
```

Reply:
Yeah, I wanted to do that but I just wasn't sure it would be alright and forgot to ask. Thanks for the clarification.